### PR TITLE
js1-week3-homework

### PR DESCRIPTION
**Pull request by mehmetekinci**

### CODE REVIEW

File: Week3/script.js [hunk 1]
```diff
@@ -0,0 +1,273 @@
+/* *** String! *** */
+
+// 1.1
```
Comment (marcianoviereck) [inline Week3/script.js:3]: Could you split up your homework in separate files?

See this:

> For all the following exercises create a new .js file.

File: Week3/script.js [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+console.log(myString);
+
+// 1.2
+
+console.log(myString.length);
+
+// 1.3 & 1.4
+
+console.log(myString.replace(/,/g, " "));
```
Comment (marcianoviereck) [inline Week3/script.js]: Nice solution!

File: Week3/script.js [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+// 2.9
+
+console.log(favoriteAnimals.indexOf("meerkat"));
+
+// 2.10
+
+console.log("The item you are looking for is at index: 1");
+
+        /* *** More JavaScript *** */
```
Comment (marcianoviereck) [inline Week3/script.js]: I think you might have forgotten to remove this comment?

File: Week3/script.js [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+function threeArguments(a, b, c) {
+  var sum = a + b + c;
+
+  return sum;
+}
+
+        // 2
+
+function colorCar(car) {
+  return car;
```
Comment (marcianoviereck) [inline Week3/script.js]: Looks good! But I think the function itself should `console.log("a red car")`. Now the function only returns the variable to you pass into it. But it should use the value you pass into it to `console.log` the correct value.

File: Week3/script.js [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+let car = "a red car";
+
+console.log(colorCar(car));
+
+        // 3
+
+const person = {
+  name: "Mehmet Ekinci",
+  gender: "M",
+  adres() {
```
Comment (marcianoviereck) [inline Week3/script.js]: I don't think this works as a property, did you try to run this part of the code?

File: Week3/script.js [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    console.log("Koemeershoeve");
+  }
+};
+console.log(person);
+
+        // 4
+
+function vehicleType(color,code) {
```
Comment (marcianoviereck) [inline Week3/script.js]: looks nice!

File: Week3/script.js [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+    type = " car";
+  } if (code === 2) {
+    type = " motorbike";
+  } if (age >= 2) {
+    condition = " used";
+  } if ((age >= 0) && (age <= 1)) {
+    condition = " new";
+  }
+  
+  console.log("a " + color + condition + type);
```
Comment (marcianoviereck) [inline Week3/script.js]: This looks nice, but I would recommend to add some more spaces and use `else if`. Like this:

```
  if (code === 1) {
    type = " car";
  } else if (code === 2) {
    type = " motorbike";
  } 
  if ((age >= 0) && (age <= 1)) {
    condition = " new";
  } else if (age >= 2) {
    condition = " used";
  } 
```

This makes it easier to read (and the logic is easier to follow:))

File: Week3/script.js [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+
+}
+vehicle("blue",1,5);
+
+        // 7
+
+let vehicles = ["motorbike", "caravan", "bike", "car", ]
```
Comment (marcianoviereck) [inline Week3/script.js]: Does it work with the `,` at the end? (and don't forget the `;` ;))

File: Week3/script.js [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+
+        // 8
+
+console.log(vehicles[2]);
+
+        // 9
+
+function newVehicle(color,code,age) {
```
Comment (marcianoviereck) [inline Week3/script.js]: I know it's not clear from the assignment, but the `code` should be used to get the value from the array.

File: Week3/script.js [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+        // 10
+let advertisement = "Amazing Joe's Garage, we service "
+function reAdvertisement(advertisement, vehicles){
+
+  for( let i = 0; i < vehicles.length ;i++){
+    const lastElement = vehicles[vehicles.length-1];
+    const beforLastElement = vehicles[vehicles.length-2];
+    const currentElement = vehicles[i];
+
+    if (currentElement === lastElement) {
```
Comment (marcianoviereck) [inline Week3/script.js]: Ah nice!

This looks good, you can also use the `indexes` though and then compare the `index` numbers. So then it would look like this:

```
for( let i = 0; i < vehicles.length ;i++){
    if (i=== vehicles.length - 1) {
     advertisement = advertisement + currentElement + ".";
    } else if (i=== vehicles.length - 2) {
      advertisement = advertisement + currentElement + " and ";
    } else {
      advertisement = advertisement + currentElement + ", "
    }
```

File: Week3/script.js [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+console.log(reAdvertisement(advertisement,vehicles));
+
+        // 11
+
+// If you add more vehicle to array, you will be add to advertisement.
+function newFunction(){
+vehicles.push("truck");
+ return console.log(reAdvertisement(advertisement,vehicles))
+}
+newFunction();
```
Comment (marcianoviereck) [inline Week3/script.js:182]: nice!

File: Week3/script.js [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+
+        // 12
+
+let newObject = {};
+
+        // 13
+
+let hackYourFuture = new Object();
+
+hackYourFuture.teachers = "Philipp, Rob, Unmesh, Bonan, Yash";
```
Comment (marcianoviereck) [inline Week3/script.js]: It's good practice to create an `array` object for multiple values, instead of a string with `commas`.
So could you change this into an array:)?

File: Week3/script.js [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+
+console.log(hackYourFuture);
+
+
+
+        // 14
+
+hackYourFuture.langs = "HTML, CSS, CLI, GIT-Bash, JS";
```
Comment (marcianoviereck) [inline Week3/script.js]: Same here

File: Week3/script.js [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+
+const o1 = { foo: "bar" };
+const o2 = { foo: "bar" };
+const o3 = o2;
+console.log(o2.foo)
+console.log(o3)
+
+console.log(o1 == o2);  // false
+console.log(o2 == o3);  // true
+console.log(o3 == 01);  // false
```
Comment (marcianoviereck) [inline Week3/script.js]: I think you wrote `01` instead of `o1` (zero instead of o)

Comment (marcianoviereck) [reply Week3/script.js]: Can you also change that?

**Review by marcianoviereck — CHANGES_REQUESTED** [review]
The homework looks really nice overall:) I do have some review comments here and there. But overall the functions are really nice, keep it up!

**Comment by yash-kapila** [comment]
@mehmetekinci could you please have a look at the review comments and try to implement the requested changes? 🙂

File: Week3/script.js (hunk 1 above)
Comment (mehmetekinci) [reply Week3/script.js:3]: oo I'm sorry for that. Should I do new pull request?

**Comment by mehmetekinci** [comment]
@marcianoviereck  Thank you for your help and correction. I corrected all.

**Comment by mehmetekinci** [comment]
@yash-kapila I did it. thank you for reminding :)

File: Week3/script.js (hunk 10 above)
Comment (mehmetekinci) [reply Week3/script.js]: @marcianoviereck  Thank you :)  I actually knew that way but I tried to make from different way. Which way is the best?

File: Week3/script.js (hunk 5 above)
Comment (mehmetekinci) [reply Week3/script.js]: after your feedback I tried another thing. I hope it works.

File: Week3/script.js (hunk 10 above)
Comment (marcianoviereck) [reply Week3/script.js]: Ah, both ways are fine:) I think the one I suggested is a bit easier (and less error-prone with the comparisons). So it's a `safer` option to compare numbers instead of whole objects:)

File: Week3/script.js [hunk 15]
```diff
@@ -204,15 +203,15 @@ let newObject = {};
...
+hackYourFuture.teachers = ['Philipp', 'Rob', 'Unmesh', 'Bonan', 'Yash'];
 
 console.log(hackYourFuture);
 
 
 
         // 14
 
-hackYourFuture.langs = "HTML, CSS, CLI, GIT-Bash, JS";
+hackYourFuture.langs = ['HTML', 'CSS', 'CLI', 'GIT-Bash', 'JS'];
```
Comment (marcianoviereck) [inline Week3/script.js]: Ah, i made a mistake in my first homework-review! I said that using an `array` is a better option and I think it is a better option. But now it's not clear which teacher has taught which class(es). Do you know how you can change these code so that it is clearer which teacher taught which classes(Hint: you can use object properties and arrays.

If you don't understand what I mean, please let me know! Then I will try to clarify more.

File: Week3/script.js (hunk 1 above)
Comment (marcianoviereck) [reply Week3/script.js:3]: No, no problem!

File: Week3/script.js [hunk 16]
```diff
@@ -72,24 +72,22 @@ function threeArguments(a, b, c) {
 
         // 2
 
-function colorCar(car) {
-  return car;
+function colorCar() {
```
Comment (marcianoviereck) [inline Week3/script.js]: Ah, now the function `always` returns `a red car`. Can you change the code so that the color that it will print out can be passed in the function. So something like this:

```
function  colorCar(color) {
 //do something with the color to print "a _ car"
}
```
```

File: Week3/script.js [hunk 17]
```diff
@@ -72,24 +72,22 @@ function threeArguments(a, b, c) {
...
-  gender: "M",
-  adres() {
-    console.log("Koemeershoeve");
-  }
+  name: 'Mehmet',
 };
-console.log(person);
+
+function personInfo() {
+  return `His ${Object.keys(person)} is ${person.name}`;
```
Comment (marcianoviereck) [inline Week3/script.js]: Actually I think just `console.log(person)` is fine. There is a nicer way of doing it, but for now this is fine:).

To read more about ways to do this, you can read [this](https://zellwk.com/blog/looping-through-js-objects/) (and if you want, you can change your code;)

File: Week3/script.js [hunk 18]
```diff
@@ -120,11 +118,12 @@ function vehicle(color,code,age) {
 
   if (code === 1) {
     type = " car";
-  } if (code === 2) {
+  } els if (code === 2) {
```
Comment (marcianoviereck) [inline Week3/script.js]: Ah almost, there is a typo here. Can you take a look at this again?

File: Week3/script.js [hunk 19]
```diff
@@ -135,7 +134,7 @@ vehicle("blue",1,5);
 
         // 7
 
-let vehicles = ["motorbike", "caravan", "bike", "car", ]
+let vehicles = [" motorbike", " caravan", " bike", " car"];
```
Comment (marcianoviereck) [inline Week3/script.js]: Ah, why did you add the spaces here? I think it was fine before (except for the `,`)

File: Week3/script.js [hunk 20]
```diff
@@ -146,13 +145,13 @@ console.log(vehicles[2]);
...
+    type = vehicles[0];
   } if (code === 2) {
-    type = " caravan";
+    type = vehicles[1];
   }  if (code === 3) {
-    type = " bike";
+    type = vehicles[2];
   } if (code === 4) {
-    type = " car";
+    type = vehicles[3];
```
Comment (marcianoviereck) [inline Week3/script.js]: Ah, do you see a `pattern` here? Now you changed the code to this:

```
  if (code === 1) {
    type = vehicles[0];
  }
  if (code === 2) {
    type = vehicles[1];
}
..
```

The `index` of the array that you use, is always `1` value lower than the value checked in your `if-statement`. Do you think you can change the code so that you can use the value in `code` and use it as an `index` for the list?

**Review by marcianoviereck — CHANGES_REQUESTED** [review]
It's starting to look better! I do have some change requests:) If anything is not clear, feel free to slack me!

**Comment by mehmetekinci** [comment]
@marcianoviereck I trried to correct it again. I hope,I succeed it :)

File: Week3/script.js [hunk 21]
```diff
@@ -54,192 +52,201 @@ console.log(favoriteAnimals);
...
+const personInfo = {
   name: 'Mehmet',
 };
 
-function personInfo() {
-  return `His ${Object.keys(person)} is ${person.name}`;
+const entries = Object.entries(personInfo);
+
+for (const [personInfo, name] of entries) {
+  console.log(`His ${personInfo} is ${name}`);
```
Comment (marcianoviereck) [inline Week3/script.js:86]: Nice!

File: Week3/script.js [hunk 22]
```diff
@@ -54,192 +52,201 @@ console.log(favoriteAnimals);
...
-  } if (code === 4) {
-    type = vehicles[3];
-  } if (age >=2) {
-    condition = " used";
-  } if ((age >= 0) && (age <=1)) {
-    condition = " new";
+    type = vehicles[code - 1];
+  }
+  if (code === 2) {
+    type = vehicles[code - 1];
```
Comment (marcianoviereck) [inline Week3/script.js]: Ah, it's looking better! You can also do this:

```
 type = vehicles[code - 1];
```

Then you don't need all those 
```
 if (code === 2) {
}
if(code === 3) {
}
```

etc. So could you change that aswell?

File: Week3/script.js [hunk 23]
```diff
@@ -54,192 +52,201 @@ console.log(favoriteAnimals);
...
+  }
+  if (code === 3) {
+    type = vehicles[code - 1];
+  }
+  if (code === 4) {
+    type = vehicles[code - 1];
+  }
+  if (age >= 2) {
```
Comment (marcianoviereck) [inline Week3/script.js:147]: Nice! This looks nice

File: Week3/script.js [hunk 24]
```diff
@@ -54,192 +52,201 @@ console.log(favoriteAnimals);
...
+
+// 10
+let advertisement = "Amazing Joe's Garage, we service ";
+function reAdvertisement(advertisement, vehicles) {
+  for (let i = 0; i < vehicles.length; i++) {
+    // const lastElement = vehicles[vehicles.length - 1];
+    // const beforLastElement = vehicles[vehicles.length - 2];
+    // const currentElement = vehicles[i];
+
+    if (i === vehicles.length - 1) {
```
Comment (marcianoviereck) [inline Week3/script.js:163]: Ah nice! This looks very good, I think you forgot to remove the `commented` code on line 171, 172 and 173

File: Week3/script.js [hunk 25]
```diff
@@ -54,192 +52,201 @@ console.log(favoriteAnimals);
...
 console.log(hackYourFuture);
 
+// 14
 
-
-        // 14
-
-hackYourFuture.langs = ['HTML', 'CSS', 'CLI', 'GIT-Bash', 'JS'];
+hackYourFuture.teacher1.language = ['HTML'];
+hackYourFuture.teacher2.language = ['CSS'];
```
Comment (marcianoviereck) [inline Week3/script.js:208]: Ah very good! 😎

File: Week3/script.js [hunk 26]
```diff
@@ -0,0 +1,277 @@
...
+console.log(o1 == o2); // true
+console.log(o2 == o3); // false
+console.log(o3 == o1); // false
+console.log(o2 == o1); // false
+
+// when we change o2 whit o3, giving syntax error
+
+// Does the order that you assign (o3 = o2 or o2 = o3) matter?
+
+// sure. when you use o2 = o3, o2 declares twice. giving syntax error.
```
Comment (marcianoviereck) [inline Week3/script.js]: Ah could you show this in code? I understand that you explained it in comments, but could you explain it using code?

**Review by marcianoviereck — CHANGES_REQUESTED** [review]
Most of the changes look really nice! I do have some small feedback left (maybe some of it I forgot the first time, sorry for that)

**Comment by mehmetekinci** [comment]
@marcianoviereck hi Marciano I am sorry for late respond. I made some changes. I learn a loth with your feedbacks thank you for all.

File: Week3/script.js [hunk 27]
```diff
@@ -250,24 +257,21 @@ const o1 = o2;
...
 
 // when we change o2 whit o3, giving syntax error
 
 // Does the order that you assign (o3 = o2 or o2 = o3) matter?
 
-// sure. when you use o2 = o3, o2 declares twice. giving syntax error.  
-
-
-
+// sure. when you use o2 = o3, o2 declares twice. giving syntax error.
```
Comment (marcianoviereck) [inline Week3/script.js:257]: Ah yeah true, but in the assignment the code is as follows:

```
let o1 = { foo: "bar" };
let o2 = { foo: "bar" };
let o3 = o2;
```

And you did `const` then it's true that an error will occur :) but if you use `let`, will the order that you assign o3 = o2 or o2 = o3 matter?

And if you change o2, will it change o3? 
And if you change o3, will it change o2?

**Review by marcianoviereck — CHANGES_REQUESTED** [review]
I only have one more change request :) the rest looks good!
